README
raidrecover tool is a fork of raidextract tool http://www.chiark.greenend.org.uk/~peterb/linux/raidextract/



CHANGELOG
=== 20120121 ===
Initial code from raidextract.
Patch from http://blogmal.42.org/tidbits/raid-recovery-II.story applied:

--- raidextract.c 2008-07-26 11:33:53.000000000 +0200
+++ raidextract-new.c 2010-06-28 13:49:54.000000000 +0200
@@ -316,8 +316,10 @@
  int paritydisk=(stripe / (disks-1) + rotate) % disks;
  int len=stripesize-offset;
  int bytes;
+ int ndisk;
  char *ptr;
 
+ ndisk=(disk-paritydisk+3)%disks;
  if (!noparity && paritydisk <= disk) disk++;
  if (len>raidlen) len=raidlen;
  if (winoffset+len > datasize) len=datasize-winoffset;
@@ -337,7 +339,7 @@
  raidstart+=len;
  raidlen-=len;
 
- ptr=window[disk][windowalt]+winoffset;
+ ptr=window[ndisk][windowalt]+winoffset;
  while (len)
  {
    bytes=write(STDOUT_FILENO, ptr, len);

======


raidextract-0.2.2/lib.c
#include "lib.h"
#include "util.h"

char outbuf[BUFSIZ], errbuf[BUFSIZ];
FILE stdoutobj={0,1,0,outbuf}, stderrobj={0,2,1,errbuf};
FILE *stdout=&stdoutobj, *stderr=&stderrobj;

int errno;

void fflush(FILE *file)
{
	char *ptr=file->buf;
	int len=file->tail;
	int fd=file->fd;
	
	while (len)
	{
		int bytes=write(fd, ptr, len);
		if (bytes==-1 && errno==EINTR) continue;
		if (bytes==-1) die("write (fd %i): %s\n", fd, strerror(errno));
		len-=bytes, ptr+=bytes;			
	}
	file->tail=0;
}

int fclose(FILE *file)
{
	if (close(file->fd)) die("close %s\n", strerror(errno));
	return 0;
}

static inline void putstring(FILE *file, const char *s)
{
	int len=strlen(s);
	int tlen=BUFSIZ-file->tail;

	if (tlen>len) tlen=len;
	memcpy(file->buf+file->tail, s, tlen);
	file->tail+=tlen;
	len-=tlen;
	s+=tlen;
	if (file->tail==BUFSIZ) fflush(file);

	while (len>=BUFSIZ)
	{
		memcpy(file->buf+file->tail, s, BUFSIZ);
		file->tail=BUFSIZ;
		len-=BUFSIZ;
		s+=BUFSIZ;
		fflush(file);
	}

	memcpy(file->buf, s, len);
	file->tail+=len;
}

static inline void putulonglong(FILE *file, unsigned long long lli)
{
	char buf[256];
	char *ptr=buf+256;
	
	*--ptr=0;
	do
	{
		*--ptr = '0' + (lli % 10);
		lli /= 10;		
	} while (lli);
	putstring(file, ptr);
}

static inline void putlonglong(FILE *file, long long lli)
{
	char buf[256];
	char *ptr=buf+256;
	int sign=0;
	
	if (lli<0) sign=1,lli=-lli;
	*--ptr=0;
	do
	{
		*--ptr = '0' + (lli % 10);
		lli /= 10;		
	} while (lli);
	if (sign) *ptr-- = '-';
	putstring(file, ptr);
}

void vfprintf(FILE *file, const char *format, va_list ap)
{
	int fmt=0, lfmt=0;
	char c;
	const char *s, *oformat=format;
	long long lli=0;

	while ((c=*format++))
	{
		if (fmt)
		{
			switch(c)
			{
			case 'l':
				lfmt++;
				break;
			case 'c':
				c=va_arg(ap, int);
				putc(c, file);
				fmt=0;
				break;
			case 's':
				s=va_arg(ap, const char *);
				putstring(file, s);
				fmt=0;
				break;
			case 'd':
			case 'i':
				switch(lfmt)
				{
				case 0:
					lli=va_arg(ap, int);
					break;
				case 1:
					lli=va_arg(ap, long);
					break;
				case 2:
					lli=va_arg(ap, long long);
					break;
				default:
					die("Bad integer format in %s\n",
					    format);
				}
				putlonglong(file, lli);
				fmt=0;
				break;
			case 'u':
				switch(lfmt)
				{
				case 0:
					lli=va_arg(ap, int);
					break;
				case 1:
					lli=va_arg(ap, long);
					break;
				case 2:
					lli=va_arg(ap, long long);
					break;
				default:
					die("Bad integer format in %s\n",
					    format);
				}
				putulonglong(file, lli);
				fmt=0;
				break;
			default:
				die("Bad format char %c in %s\n", c, oformat);
			}
		}
		else
		{
			if (c=='%')
				fmt=1, lfmt=0;
			else
				putc(c, file);
		}
	}
	if (file->autoflush) fflush(file);
}

void fprintf(FILE *file, const char *format, ...)
{
        va_list ap;
        va_start(ap, format);
        vfprintf(file, format, ap);
        va_end(ap);
}

void exit(int status)
{
	fflush(stdout);
	fflush(stderr);
	_exit(status);
}

extern void *mmap(void *start, size_t length, int prot, int flags, int fd,
		  off_t offset)
{
	return (void *)mmap2((unsigned long)start, length, prot, flags, fd, offset);
}


const char *strerror(int err)
{
	const char *errmsg[125]=
		{
			"Success",
			"Operation not permitted",
			"No such file or directory",
			"No such process",
			"Interrupted system call",
			"I/O error",
			"No such device or address",
			"Arg list too long",
			"Exec format error",
			"Bad file number",
			"No child processes",
			"Try again",
			"Out of memory",
			"Permission denied",
			"Bad address",
			"Block device required",
			"Device or resource busy",
			"File exists",
			"Cross-device link",
			"No such device",
			"Not a directory",
			"Is a directory",
			"Invalid argument",
			"File table overflow",
			"Too many open files",
			"Not a typewriter",
			"Text file busy",
			"File too large",
			"No space left on device",
			"Illegal seek",
			"Read-only file system",
			"Too many links",
			"Broken pipe",
			"Math argument out of domain of func",
			"Math result not representable",
			"Resource deadlock would occur",
			"File name too long",
			"No record locks available",
			"Function not implemented",
			"Directory not empty",
			"Too many symbolic links encountered",
			"Unknown error 41",
			"No message of desired type",
			"Identifier removed",
			"Channel number out of range",
			"Level 2 not synchronized",
			"Level 3 halted",
			"Level 3 reset",
			"Link number out of range",
			"Protocol driver not attached",
			"No CSI structure available",
			"Level 2 halted",
			"Invalid exchange",
			"Invalid request descriptor",
			"Exchange full",
			"No anode",
			"Invalid request code",
			"Invalid slot",
			"Unknown error 58",
			"Bad font file format",
			"Device not a stream",
			"No data available",
			"Timer expired",
			"Out of streams resources",
			"Machine is not on the network",
			"Package not installed",
			"Object is remote",
			"Link has been severed",
			"Advertise error",
			"Srmount error",
			"Communication error on send",
			"Protocol error",
			"Multihop attempted",
			"RFS specific error",
			"Not a data message",
			"Value too large for defined data type",
			"Name not unique on network",
			"File descriptor in bad state",
			"Remote address changed",
			"Can not access a needed shared library",
			"Accessing a corrupted shared library",
			".lib section in a.out corrupted",
			"Attempting to link in too many shared libraries",
			"Cannot exec a shared library directly",
			"Illegal byte sequence",
			"Interrupted system call should be restarted",
			"Streams pipe error",
			"Too many users",
			"Socket operation on non-socket",
			"Destination address required",
			"Message too long",
			"Protocol wrong type for socket",
			"Protocol not available",
			"Protocol not supported",
			"Socket type not supported",
			"Operation not supported on transport endpoint",
			"Protocol family not supported",
			"Address family not supported by protocol",
			"Address already in use",
			"Cannot assign requested address",
			"Network is down",
			"Network is unreachable",
			"Network dropped connection because of reset",
			"Software caused connection abort",
			"Connection reset by peer",
			"No buffer space available",
			"Transport endpoint is already connected",
			"Transport endpoint is not connected",
			"Cannot send after transport endpoint shutdown",
			"Too many references: cannot splice",
			"Connection timed out",
			"Connection refused",
			"Host is down",
			"No route to host",
			"Operation already in progress",
			"Operation now in progress",
			"Stale NFS file handle",
			"Structure needs cleaning",
			"Not a XENIX named type file",
			"No XENIX semaphores available",
			"Is a named type file",
			"Remote I/O error",
			"Quota exceeded",
			"No medium found",
			"Wrong medium type",
		};
	
	if (err<0) return "invalid error";
	if (err>=125) return "unknown error";
	return errmsg[err];
}

int strcmp(const char *s1, const char *s2)
{
	char c, d;
	while ((c=*s1)==(d=*s2) && *s1) s1++,s2++;
	return c-d;
}

size_t strlen(const char *s)
{
	const char *p=s;
	while (*p) p++;
	return p-s;
}

void *memcpy(void *dest, const void *src, size_t n)
{
	char *d=(char *)dest;
	const char *s=(const char *)src;
	while (n--) *d++=*s++;
	return dest;
}

char *strrchr(const char *s, int c)
{
	unsigned char *p=(unsigned char *)s+strlen(s);
	while (p-- > (unsigned char *)s)
		if (*p==c) return p;
	return NULL;	
}

int atoi(const char *nptr)
{
	char c;
	int i=0;
	int sign=0;
	if (*nptr=='-')
		sign=1,nptr++;
	else if (*nptr=='+')
		nptr++;		
	while ((c=*nptr++))
	{
		if (c<'0' || c>'9') break;
		i=i*10+(c-'0');
	}
	return sign? -i:i;
}

long long atoll(const char *nptr)
{
	char c;
	long long i=0;
	int sign=0;
	if (*nptr=='-')
		sign=1,nptr++;
	else if (*nptr=='+')
		nptr++;		
	while ((c=*nptr++))
	{
		if (c<'0' || c>'9') break;
		i=i*10+(c-'0');
	}
	return sign? -i:i;
}

int _main(int argc, char *ptr)
{
	char *argv[argc+1]; /* gcc / C99 feature */
	int i=0;

	while (i<argc)
	{
		argv[i++]=ptr;
		while (*ptr++);
	}
	argv[i]=0;
	return main(argc, argv);;
}


raidextract-0.2.2/lib.h
#ifndef LIB_H
#define LIB_H

typedef long long off_t;
typedef unsigned long size_t;
typedef int uint32_t;

#include <stdarg.h>

#define BUFSIZ 4096

typedef struct
{
	int tail, fd, autoflush;
	char *buf;
} FILE;

extern FILE *stdout, *stderr;

extern void fprintf(FILE *file, const char *format, ...);
extern void vfprintf(FILE *stream, const char *format, va_list ap);
extern int atoi(const char *nptr);
extern long long atoll(const char *nptr);

extern int errno;

#define __syscall_return(type, res) \
do { \
	if ((unsigned long)(res) >= (unsigned long)(-125)) { \
		errno = -(res); \
		res = -1; \
	} \
	return (type) (res); \
} while (0)
#define _syscall0(type,name) \
type name(void) \
{ \
long __res; \
__asm__ volatile ("int $0x80" \
	: "=a" (__res) \
	: "0" (__NR_##name)); \
__syscall_return(type,__res); \
}
#define _syscall1mem(type,name,type1,arg1) \
type name(type1 arg1) \
{ \
long __res; \
__asm__ volatile ("int $0x80" \
	: "=a" (__res) \
	: "0" (__NR_##name),"b" ((long)(arg1)) \
        ); \
__syscall_return(type,__res); \
}
#define _syscall1(type,name,type1,arg1) \
type name(type1 arg1) \
{ \
long __res; \
__asm__ volatile ("int $0x80" \
	: "=a" (__res) \
	: "0" (__NR_##name),"b" ((long)(arg1)) \
	: "memory" ); \
__syscall_return(type,__res); \
}
#define _syscall2mem(type,name,type1,arg1,type2,arg2) \
type name(type1 arg1,type2 arg2) \
{ \
long __res; \
__asm__ volatile ("int $0x80" \
	: "=a" (__res) \
	: "0" (__NR_##name),"b" ((long)(arg1)),"c" ((long)(arg2)) \
	: "memory" ); \
__syscall_return(type,__res); \
}
#define _syscall2(type,name,type1,arg1,type2,arg2) \
type name(type1 arg1,type2 arg2) \
{ \
long __res; \
__asm__ volatile ("int $0x80" \
	: "=a" (__res) \
	: "0" (__NR_##name),"b" ((long)(arg1)),"c" ((long)(arg2)) \
        ); \
__syscall_return(type,__res); \
}
#define _syscall3mem(type,name,type1,arg1,type2,arg2,type3,arg3) \
type name(type1 arg1,type2 arg2, type3 arg3) \
{ \
long __res; \
__asm__ volatile ("int $0x80" \
	: "=a" (__res) \
	: "0" (__NR_##name),"b" ((long)(arg1)),"c" ((long)(arg2)), \
		  "d" ((long)(arg3)) \
	: "memory" ); \
__syscall_return(type,__res); \
}
#define _syscall3(type,name,type1,arg1,type2,arg2,type3,argc) \
type name(type1 arg1,type2 arg2, type3 arg3) \
{ \
long __res; \
__asm__ volatile ("int $0x80" \
	: "=a" (__res) \
	: "0" (__NR_##name),"b" ((long)(arg1)),"c" ((long)(arg2)), \
		  "d" ((long)(arg3)) \
        ); \
__syscall_return(type,__res); \
}
#define _syscall6mem(type,name,type1,arg1,type2,arg2,type3,arg3,type4,arg4, \
	  type5,arg5,type6,arg6) \
type name (type1 arg1,type2 arg2,type3 arg3,type4 arg4,type5 arg5,type6 arg6) \
{ \
long __res; \
__asm__ volatile ("push %%ebp ; movl %%eax,%%ebp ; movl %1,%%eax ; int $0x80 ; pop %%ebp" \
	: "=a" (__res) \
	: "i" (__NR_##name),"b" ((long)(arg1)),"c" ((long)(arg2)), \
	  "d" ((long)(arg3)),"S" ((long)(arg4)),"D" ((long)(arg5)), \
	  "0" ((long)(arg6)) \
	: "memory"); \
__syscall_return(type,__res); \
}

#define __NR__exit	1
#define __NR_fork	2
#define __NR_read	3
#define __NR_write	4
#define __NR_open	5
#define __NR_close	6
#define __NR_pipe	42
#define __NR_munmap	91
#define __NR__llseek	140
#define __NR_msync	144
#define __NR_mmap2	192

static inline _syscall1(int,_exit,int,exitcode)
static inline _syscall0(int,fork)
static inline _syscall3mem(int,read,int,fd,void *,buf,size_t,count)
static inline _syscall3mem(int,write,int,fd,const void *,buf,size_t,count)
static inline _syscall3(off_t,_llseek,int,fd,off_t,offset,int,count)
static inline _syscall3(int,open,const char *,file,int,flag,int,mode)
static inline _syscall1(int,close,int,fd)
static inline _syscall6mem(int,mmap2,unsigned long,addr,unsigned long,len,
			unsigned long,prot,unsigned long,flags,
			unsigned long,fd,unsigned long,pgoff)
static inline _syscall2mem(int,munmap,void *,start,size_t,length)
static inline _syscall3mem(int,msync,const void *,start,size_t,length,int,flags)
static inline _syscall1mem(int,pipe,int *,fds)

extern void exit(int status);
extern const char *strerror(int err);
extern void *mmap(void *start, size_t length, int prot, int flags, int fd,
		  off_t offset);

#define	EINTR		 4	/* Interrupted system call */
#define STDOUT_FILENO    1
#define PROT_READ	0x1		/* page can be read */
#define PROT_WRITE	0x2		/* page can be written */
#define MAP_SHARED	0x01		/* Share changes */
#define MAP_ANONYMOUS	0x20		/* don't use a file */
#define MS_INVALIDATE	2		/* invalidate the caches */
#define MS_SYNC		4		/* synchronous memory sync */
#define MAP_FAILED	((void *)-1)
#define O_RDONLY	     00
#define O_LARGEFILE	0100000
#define NULL ((void *)0)
#define SEEK_SET	0	/* Seek from beginning of file.  */

extern int strcmp(const char *s1, const char *s2);
extern size_t strlen(const char *s);
extern void *memcpy(void *dest, const void *src, size_t n);
extern char *strrchr(const char *s, int c);

extern int _main(int argc, char *ptr);
extern int main(int argc, char **argv);

/* We halt immediately when an error occurs - there is no error flag */
static inline int ferror(FILE *file) { file=file; return 0; } 
extern int fclose(FILE *file);

#define lseek64 _llseek
#define putchar(c) putc(c, stdout)
#define printf(fmt, ...) fprintf(stdout, fmt, ## __VA_ARGS__)

extern void fflush(FILE *file);
static inline void putc(int c, FILE *file)
{
	file->buf[file->tail++]=c;
	if (file->tail==BUFSIZ) fflush(file);
}

#endif


raidextract-0.2.2/parityinfo.c
/* Comile with -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 */

#ifdef LINUXSTATIC
#include "lib.h"
#else
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include <errno.h>
#include <unistd.h>
#include <fcntl.h>
#include <stdint.h>
#include <sys/mman.h>
#endif

#include "util.h"

#define MAXDISKS 12
#define DEFSTRIPE 64
#define DEFWINDOW 128

const char *diskname[MAXDISKS], *progname="parityinfo";
int diskfd[MAXDISKS], request[MAXDISKS], reply[MAXDISKS];
off_t windowstart, windowend, windowcyl, stripecyl,
	raidstart=0, raidlen=0;
int disks=0, diskstart, datasize=0, rotate=0,
	stripesize=DEFSTRIPE*1024, windowsize=DEFWINDOW*1024;

void usage(int status)
{
	fprintf(status? stderr:stdout, "Usage: %s [options] disks...\n"
		"        --stripe <n>"
		"        Set stripe size in K (default: %i)\n"
		"        --window <n>"
		"        Set I/O window size in K (default: %i)\n"
		"        --start <n> "
		"        Skip to given output position in bytes (default: 0)\n"
		"        --length <n>"
		"        Length of output (no default)\n"
		"        --rotate <n>"
		"        Rotate parity disk position (default: 0)\n",
		progname, DEFSTRIPE, DEFWINDOW);
	exit(status);
}

void parseopts(int argc, char *argv[])
{
	int i, optparsing=1;
	
	if (argc) setprogname(argv[0]);
	
	if (argc<=1) usage(1);

	for (i=1; i<argc; i++)
	{
		if (optparsing && argv[i][0] == '-')
		{
			if (!strcmp(argv[i], "--"))
			{
				optparsing=0;
				continue;
			}
			if (!strcmp(argv[i], "--help"))
				usage(0);
			if (!strcmp(argv[i], "--stripe"))
			{
				if (++i==argc) 
					die("--stripe takes an argument\n");
				stripesize=atoi(argv[i])*1024;
				if (stripesize <= 0) 
					die("Invalid stripe size\n");
				continue;
			}
			if (!strcmp(argv[i], "--window"))
			{
				if (++i==argc) 
					die("--window takes an argument\n");
				windowsize=atoi(argv[i])*1024;
				if (windowsize <= 0) 
					die("Invalid window size\n");
				continue;
			}
			if (!strcmp(argv[i], "--start"))
			{
				if (++i==argc) 
					die("--start takes an argument\n");
				raidstart=atoll(argv[i]);
				if (raidstart < 0) 
					die("Invalid start position\n");
				continue;
			}
			if (!strcmp(argv[i], "--length"))
			{
				if (++i==argc) 
					die("--length takes an argument\n");
				raidlen=atoll(argv[i]);
				if (raidlen <= 0) 
					die("Invalid length\n");
				continue;
			}
			if (!strcmp(argv[i], "--rotate"))
			{
				if (++i==argc) 
					die("--rotate takes an argument\n");
				rotate=atoi(argv[i]);
				continue;
			}
			die("Unknown option %s\n", argv[i]);
		}
		if (disks==MAXDISKS) die("Too many disks -- increase MAXDISKS\n");
		diskname[disks++]=argv[i];
	}
	if (disks<2) die("Need at least two disk names\n");
	if (windowsize % stripesize) 
		die("Window size must be a multiple of stripe size\n");
	if (raidlen <= 0) die("Need to specify mandatory argument: --length\n");
}

int getnextwindow(void)
{
	windowstart += datasize;
	datasize = windowsize;
	windowend = windowstart + windowsize-1;
	printf("Reading from all disks, position %lli to %lli "
	       "(stripe %lli to %lli)\n", 
	       windowstart, windowend, 
	       windowstart/stripesize, windowend/stripesize);
	return windowsize;
}

void printpartial(void)
{
	off_t windowoffset = raidstart - windowstart*(disks-1);
	int stripe=raidstart/stripesize;
	int offset=raidstart%stripesize;
	off_t winoffset = (windowoffset/stripecyl)*stripesize+offset;
	int disk=stripe % (disks-1);
	int paritydisk=(stripe / (disks-1) + rotate) % disks;
	int len=stripesize-offset;

	if (paritydisk <= disk) disk++;
	if (len>raidlen) len=raidlen;
	if (winoffset+len > datasize) len=datasize-winoffset;

	printf("\nOutput position %lli (stripe %i)\n",
	       raidstart, stripe);
	printf("  data disk=%i %s\n", disk, diskname[disk]);
	printf("  parity disk=%i %s\n", paritydisk, diskname[paritydisk]);
 	printf("  data %lli (stripe %lli) [%i..%i]\n", windowstart+(windowoffset/stripecyl)*stripesize, windowstart/stripesize+(windowoffset/stripecyl), offset, offset+len-1);

	raidstart+=len;
	raidlen-=len;
}

int main(int argc, char *argv[])
{
	parseopts(argc, argv);

	windowcyl = windowsize * (disks-1);
	stripecyl = stripesize * (disks-1);
	printf("Length of stripe across all disks (except parity): "
		"%llu\n", stripecyl);
	printf("Number of disks: %i\n", disks);
	windowstart = (raidstart/windowcyl) * windowsize;
	getnextwindow();

	while (raidlen)
	{
		off_t nextwindow = (raidstart/windowcyl) * windowsize;
		if (windowstart != nextwindow)
		{
			getnextwindow();
		}
		printpartial();
	}

	return 0;
}


raidextract-0.2.2/raidextract.c
/* Comile with -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 */

#ifdef LINUXSTATIC
#include "lib.h"
#else
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include <errno.h>
#include <unistd.h>
#include <fcntl.h>
#include <stdint.h>
#include <sys/mman.h>
#endif

#include "util.h"

#define MAXDISKS 12
#define DEFSTRIPE 64
#define DEFWINDOW 128

const char *diskname[MAXDISKS], *progname="raidextract";
char *window[MAXDISKS][2];
int windowalt=1, failed=-1;
int diskfd[MAXDISKS], request[MAXDISKS], reply[MAXDISKS];
off_t windowstart, windowend, windowcyl, stripecyl,
	raidstart=0, raidlen=0;
int disks=0, diskstart, datasize=0, rotate=0,
	stripesize=DEFSTRIPE*1024, windowsize=DEFWINDOW*1024;
int noparity = 0;

void usage(int status)
{
	fprintf(status? stderr:stdout, "Usage: %s [options] disks...\n"
		"        --stripe <n>"
		"        Set stripe size in K (default: %i)\n"
		"        --window <n>"
		"        Set I/O window size in K (default: %i)\n"
		"        --start <n> "
		"        Skip to given output position in bytes (default: 0)\n"
		"        --length <n>"
		"        Length of output (no default)\n"
		"        --rotate <n>"
		"        Rotate parity disk position (default: 0)\n"
		"        --noparity  "
		"        There is no parity, it's not raid5\n"
		"        --failed <n>"
		"        Ignore data on disk at index n as failed (default: none)\n",
		progname, DEFSTRIPE, DEFWINDOW);
	exit(status);
}

void diskhelper(int i, int rd, int wr)
{
	int bytes, dummy, datalen;
	while (1)
	{
		char *ptr;
		int len=windowsize;

		windowalt=!windowalt;
		ptr=window[i][windowalt];

		while (len)
		{
			bytes=read(diskfd[i], ptr, windowsize);
			if (bytes==-1 && errno==EINTR) continue;
			if (bytes==-1) 
				die("read (child %i, file %s): %s\n", 
				    i, diskname[i], strerror(errno));
			if (bytes==0) break;
			ptr+=bytes;
			len-=bytes;
		}

		datalen=windowsize-len;
		msync(window[i][windowalt], windowsize, MS_SYNC | MS_INVALIDATE);
		len=sizeof(datalen); ptr=(char *)&datalen;
		while (len)
		{
			bytes=write(wr, ptr, len);
			if (bytes==-1 && errno==EINTR) continue;
			if (bytes==-1) 
				die("write (child %i): %s\n",
				    i, strerror(errno));
			len-=bytes;
			ptr+=bytes;			
		}
		while ((bytes=read(rd, &dummy, 1))!=1)
		{
			if (bytes==-1 && errno==EINTR) continue;
			if (bytes==-1) 
				die("read (child %i): %s\n", 
				    i, strerror(errno));
			if (bytes==0) exit(0);
		}
	}
}

void parseopts(int argc, char *argv[])
{
	int i, optparsing=1;
	
	if (argc) setprogname(argv[0]);
	
	if (argc<=1) usage(1);

	for (i=1; i<argc; i++)
	{
		if (optparsing && argv[i][0] == '-')
		{
			if (!strcmp(argv[i], "--"))
			{
				optparsing=0;
				continue;
			}
			if (!strcmp(argv[i], "--help"))
				usage(0);
			if (!strcmp(argv[i], "--stripe"))
			{
				if (++i==argc) 
					die("--stripe takes an argument\n");
				stripesize=atoi(argv[i])*1024;
				if (stripesize <= 0) 
					die("Invalid stripe size\n");
				continue;
			}
			if (!strcmp(argv[i], "--window"))
			{
				if (++i==argc) 
					die("--window takes an argument\n");
				windowsize=atoi(argv[i])*1024;
				if (windowsize <= 0) 
					die("Invalid window size\n");
				continue;
			}
			if (!strcmp(argv[i], "--start"))
			{
				if (++i==argc) 
					die("--start takes an argument\n");
				raidstart=atoll(argv[i]);
				if (raidstart < 0) 
					die("Invalid start position\n");
				continue;
			}
			if (!strcmp(argv[i], "--noparity"))
			{
				noparity = 1;
				continue;
			}
			if (!strcmp(argv[i], "--length"))
			{
				if (++i==argc) 
					die("--length takes an argument\n");
				raidlen=atoll(argv[i]);
				if (raidlen <= 0) 
					die("Invalid length\n");
				continue;
			}
			if (!strcmp(argv[i], "--rotate"))
			{
				if (++i==argc) 
					die("--rotate takes an argument\n");
				rotate=atoi(argv[i]);
				continue;
			}
			if (!strcmp(argv[i], "--failed"))
			{
				if (++i==argc) 
					die("--failed takes an argument\n");
				failed=atoi(argv[i]);
				if (failed < 0) 
					die("Invalid failed disk\n");
				continue;
			}
			die("Unknown option %s\n", argv[i]);
		}
		if (disks==MAXDISKS) die("Too many disks -- increase MAXDISKS\n");
		diskname[disks++]=argv[i];
	}
	if (disks<2) die("Need at least two disk names\n");
	if (failed >= disks) die("Invalid failed disk\n");
	if (windowsize % stripesize) 
		die("Window size must be a multiple of stripe size\n");
	if (raidlen <= 0) die("Need to specify mandatory argument: --length\n");
}

void starthelpers(void)
{
	int i;

	for (i=0; i<disks; i++)
	{
		int fdsa[2], fdsb[2];
		window[i][0]=mmap(NULL, windowsize, PROT_READ | PROT_WRITE, 
			       MAP_SHARED | MAP_ANONYMOUS, -1, 0);
		if (window[i][0] == MAP_FAILED)
			die("Map failed: %s\n", strerror(errno));
		window[i][1]=mmap(NULL, windowsize, PROT_READ | PROT_WRITE, 
			       MAP_SHARED | MAP_ANONYMOUS, -1, 0);
		if (window[i][1] == MAP_FAILED)
			die("Map failed: %s\n", strerror(errno));
		if (pipe(fdsa) || pipe(fdsb))
		    die("pipe: %s\n", strerror(errno));
		request[i]=fdsa[1]; reply[i]=fdsb[0];
		if ((diskfd[i]=open(diskname[i], O_RDONLY | O_LARGEFILE, 0))==-1)
			die("open %s: %s\n", diskname[i], strerror(errno));
		if (lseek64(diskfd[i], windowstart, SEEK_SET)==-1)
		    die("lseek failed: %s\n", strerror(errno));
		switch (fork())
		{
		case -1:
			die("fork: %s\n", strerror(errno));
		case 0:
			if (i)
			{
				munmap(window[i-1][0], windowsize);
				munmap(window[i-1][1], windowsize);
				close(request[i-1]); close(reply[i-1]);
				close(diskfd[i-1]);
			}
			close(fdsa[1]); close(fdsb[0]);
			diskhelper(i, fdsa[0], fdsb[1]);
			exit(0);
		}
		close(fdsa[0]); close(fdsb[1]);
	}
}

int getnextwindow(void)
{
	int i, bytes, ret=windowsize, returned, dummy;

	for (i=0; i<disks; i++)
	{
		char *ptr=(char *)&returned;
		int len=sizeof(returned);
		while (len)
		{
			bytes=read(reply[i], ptr, len);
			if (bytes==-1 && errno==EINTR) continue;
			if (bytes==-1) die("read: %s\n", strerror(errno));
			if (!bytes) 
				die("read (parent): child %i died\n", i);
			len-=bytes;
			ptr+=bytes;
		}
		if (ret > returned) ret=returned;

		while ((bytes=write(request[i], &dummy, 1))!=1)
		{
			if (bytes==-1 && errno==EINTR) continue;
			die("write: %s\n", strerror(errno));
		}
	}

	ret=(ret/stripesize) * stripesize;
	windowstart += datasize;
	datasize = ret;
	windowend = windowstart + ret-1;
	windowalt = !windowalt;
	return ret;
}

void verifychecksum(void)
{
	int i, a;
	int lastwarn=-stripesize;

	if (noparity)
		return;

	if (failed==-1)
	{
		for (a=0; a<datasize; a+=4)
		{
			uint32_t xor=0;
			for (i=0; i<disks; i++)
				xor ^= *(uint32_t *)(window[i][windowalt]+a);
			if (xor) 
				die("Parity check failed at disk positition %lld at byte %d\n", 
				    (long long)(windowstart+a), a);
		}
	}
	else
	{
		for (a=0; a<datasize; a+=4)
		{
			uint32_t xor=0;
			for (i=0; i<disks; i++)
				xor ^= *(uint32_t *)(window[i][windowalt]+a);
			if (xor) 
			{
				if (a-lastwarn>=stripesize)
				{
					warn("Parity check failed at disk positition %lld\n", 
					     (long long)(windowstart+a));
					lastwarn=a;
				}
			}
			xor ^= *(uint32_t *)(window[failed][windowalt]+a);
			*(uint32_t *)(window[failed][windowalt]+a) = xor;
		}
		
	}
}

void printpartial(void)
{
	off_t windowoffset = raidstart - windowstart*(disks-(!noparity));
	int stripe=raidstart/stripesize;
	int offset=raidstart%stripesize;
	off_t winoffset = (windowoffset/stripecyl)*stripesize+offset;
	int disk=stripe % (disks-(!noparity));
	int paritydisk=(stripe / (disks-1) + rotate) % disks;
	int len=stripesize-offset;
	int bytes;
  int ndisk;
	char *ptr;

  ndisk=(disk-paritydisk+3)%disks;
	if (!noparity && paritydisk <= disk) disk++;
	if (len>raidlen) len=raidlen;
	if (winoffset+len > datasize) len=datasize-winoffset;
	if (!len)
		die("Input ended with %llu output bytes unwritten\n", raidlen);

/* 	fprintf(stderr, "\nraidlen=%lli\n", raidlen); */
/* 	fprintf(stderr, "raidstart=%llu\n", (long long)raidstart); */
/* 	fprintf(stderr, "windowstart=%llu\n", (long long)windowstart); */
/* 	fprintf(stderr, "windowoffset=%llu\n", (long long)windowoffset); */
/* 	fprintf(stderr, "stripe=%i\n", stripe); */
/* 	fprintf(stderr, "disk=%i\n", disk); */
/* 	fprintf(stderr, "paritydisk=%i\n", paritydisk); */
/*  	fprintf(stderr, "len=%i\n", len); */
/*  	fprintf(stderr, "print data [%lli] [%i..%i]\n", windowoffset/stripecyl, offset, offset+len-1); */

	raidstart+=len;
	raidlen-=len;

  ptr=window[ndisk][windowalt]+winoffset;
	while (len)
	{
		bytes=write(STDOUT_FILENO, ptr, len);
		if (bytes==-1 && errno==EINTR) continue;
		if (bytes==-1) die("write: %s\n", strerror(errno));
		len-=bytes;
		ptr+=bytes;			
	}
}

int main(int argc, char *argv[])
{
	parseopts(argc, argv);

	windowcyl = windowsize * (disks-(!noparity));
	stripecyl = stripesize * (disks-(!noparity));
	windowstart = (raidstart/windowcyl) * windowsize;
	starthelpers();
	getnextwindow();
	verifychecksum();

	while (raidlen)
	{
		off_t nextwindow = (raidstart/windowcyl) * windowsize;
		if (windowstart != nextwindow)
		{
			getnextwindow();
			verifychecksum();
		}
		printpartial();
	}

	return 0;
}
